### PR TITLE
Global time coordinator

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will add a global time coordinator and merge in the time request verify code as well

### CODE REVIEW

**Comment by phlptp** [comment]
@nightlark  can you review?  I think this PR is getting kind of long.  Doesn't fix everything yet, but does make progress enough that it needs to be merged.  I am going to keep working a little but will work in a different branch, see if I can resolve any more of the issues.

File: src/helics/core/GlobalTimeCoordinator.cpp [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        for (const auto& dep : dependencies) {
+            if (dep.dependent) {
+                if (dep.fedID == skipFed) {
+                    continue;
+                }
+                if (msg.action() == CMD_EXEC_REQUEST) {
+                    msg.setExtraDestData(dep.sequenceCounter);
+                }
+                msg.dest_id = dep.fedID;
+                sendMessageFunction(msg);
```
Comment (nightlark) [inline src/helics/core/GlobalTimeCoordinator.cpp:267]: Does a copy of `msg` get made, or will it be fully processed before a future iteration of the for loop alters the dest_id in the referenced message object?

File: src/helics/core/BaseTimeCoordinator.hpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    // should also be a ForwardingTimeCoordinator
+    TimeDependencies dependencies;  //!< federates which this Federate is temporally dependent on
+    /// callback used to send the messages
+    std::function<void(const ActionMessage&)> sendMessageFunction;
+    /// the identifier for inserting into the source id field of any generated messages;
+
+    GlobalFederateId mSourceId{0};
+    std::int32_t sequenceCounter{0};  //!< storage for sequence counter
+    bool noParent{true};  //!< indicator that the coordinator does not have parents
+    bool federatesOnly{false};  //!< indicator that the forwarder only operates with federates
```
[the remark this answers is not in the document]
Comment (phlptp) [reply src/helics/core/BaseTimeCoordinator.hpp:38]: at the moment only the forwarding time Coordinator uses it but it is computed in the base, so this seemed easier for the moment than refactoring out a different function for doing the computation.

File: src/helics/core/GlobalTimeCoordinator.cpp (hunk 1 above)
Comment (phlptp) [reply src/helics/core/GlobalTimeCoordinator.cpp:267]: sendMessageFunction takes  const ref, so will copy it

**Review by nightlark — APPROVED** [review]

**Comment by phlptp** [comment]
I am going to merge this, I will get the doc fixes in the next one

**Pull request merged**